Make snapshot locations configurable

acondolu opened the pull request:
**Description of the problem**
`pantry` expands snapshot synonyms (like `lts-X.Y` or `nightly-YYYY-MM-DD`) to a URL, using the hardcoded address of  commercialhaskell's git repository.

This basically happens in the functions `ltsSnapshotLocation` and `nightlySnapshotLocation` in [Types.hs](https://github.com/commercialhaskell/pantry/blob/master/src/Pantry/Types.hs). For instance, `lts-X.Y` expands to `https://raw.githubusercontent.com/commercialhaskell/stackage-snapshots/master/lts/X/Y.yaml`.

The hardcoded URL, however, prevents to use mirrors for snapshot locations, and mirrors are necessary in corporate environments (or really anywhere behind a firewall).

(Note: one could specify a custom URL instead of a snapshot synonym, but this can be undesirable for the following reason. In a corporate environment, it's common that the development environment has unrestricted access to the internet, but test/production servers have not. If the base location of snapshots is configurable, a project can still use the familiar snapshot synonyms, and just let the environment change the location accordingly.)


**Description of the proposed solution**
In the `Types.hs` file, is already present a mechanism to resolve relative (local) paths by means of the `Unresolved` type constructor. `Unresolved` means that such an object could contain relative paths, and a base path is required to resolve them to absolute paths.

My proposal is to extend `Unresolved` with an additional optional URL field, the URL being the supplied base location of the snapshots. I refactored `Unresolved` to follow the state monad pattern, and this minimised the changes. To resolve paths, one supplies an object of type `ResolveData`, which combines the old base directory, and the new base location of snapshots. 

This changes a bit the API, and that's where I wanted to hear from the maintainers.
The main change of course is in `ltsSnapshotLocation` and `nightlySnapshotLocation`, which return a `Unresolved RawSnapshotLocation`, but there are a few more altered type signatures (clearly `resolvePaths`, but also `warningsParserHelperRaw` and `warningsParserHelper`).

Also, it will be necessary to add a new entry to `PantryConfig`, so that the custom snapshot location is available to `loadSnapshotRaw` and `loadAndCompleteSnapshotRaw`, and as a consequence change the signature of `loadRawSnapshotLayer`. But first I wanted to hear from you.

**Additional context**
Closes #19.

snoyberg commented:
Alternative proposal: what about changing `ltsSnapshotLocation` and `nightlySnapshotLocation` from top level functions to configurable fields in `PantryConfig`, and allow arbitrary override functions to be set? Then, in Stack, you could add a `snapshot-location-base` config value or similar that would trigger overriding those functions.

acondolu commented:
Sure! I will try and implement your solution by today.

acondolu commented:
I just pushed the implementation of what @snoyberg proposed above. Just let me know what you think. A few points to discuss:

- An important change is that I was forced to add more constructors to `RawSnapshotLocation`. This is because when reading a `RawSnapshotLocation` from a JSON, the `PantryConfig` object is not available yet. What pantry does right now, it resolves right away snapshot synonyms to (hardcoded) URLs, but in this PR I delayed the resolution until later, when the `PantryConfig` is available. (We could actually take from stack the [datatype `SnapName`](https://github.com/commercialhaskell/stack/blob/0b2c555c3080ea374179416b00d3afa78dfbabdf/src/Stack/Types/Resolver.hs#L67))

- I removed `specialRawSnapshotLocation`, a function mapping old hardcoded URLs to snapshot synonyms. It was used to "undo" a hardcoded URL, but I think it does not make sense anymore.

- Another change, which is actually a bit spurious: I renamed `xSnaphostLocation` to `defaultXSnapshotLocation`, following the convention in `Pantry.hs`. Then I re-added the getters `xSnaphostLocation`, that just view their value from a `PantryConfig`.

snoyberg commented:
I like the idea of using the `SnapName` datatype. Then, instead of having to have two helper functions passed in, we can work with just a single `snapshotLocation :: SnapName -> RawSnapshotLocation` function, and a single extra `RSL` data constructor.

acondolu commented:
Yep, it looks much more concise now. Just let me know what else I can do to finalise this. Afterwards, I volunteer to do a PR to `stack` so to add the `snapshot-location-base` config value, and port it to the new API.

snoyberg commented:
Looks great! Before merging this, my recommendation would be to start on the Stack PR and confirm that this PR actually provides everything you need to make that one work. Then we can merge them around the same time.

acondolu commented:
A consequence of https://github.com/commercialhaskell/stack/pull/5281: since we stole `SnapName` from stack, I also moved here `parseSnapName` and `ParseSnapNameException`.

snoyberg commented on src/Pantry/Types.hs:
```diff
@@ -250,8 +252,18 @@ data PantryConfig = PantryConfig
 
+-- | Get the location of a snapshot synonym from the 'PantryConfig'.
+--
+-- @since TODO:
```
Should be 0.5.0.0, can you apply throughout this PR?

acondolu commented:
1. Yes, this is enough to make the new yaml option `snapshot-location-base` work in `stack`
2. I updated the pantry version to 0.5.0.

En passant, I added a couple of tests for `completeSnapshotLocation` when it uses `defaultSnapshotLocation`. I did not remove the previous tests about rendering LTS/Nightly names, even though they have become trivial now; I only changed their description, I'll let you decide what to do with them.

dramforever commented:
Fantastic work!
